### DISCUSSION TITLE
Overhaul Badges for Achievements

### DISCUSSION THREAD

**Discussion by CodeItQuick** [typehero/typehero#1569] [Ideas]
PR: https://github.com/typehero/typehero/pull/1573

### Community Achievements
Prioritize community achievements and use thresholds for community stuff

### Badges
1. Difficulty Completion Badge
2. Shared Solutions Up-Voted Badge

### Shared Solutions Badge
Business Rules:
1. Have shared a solution
2. Only one shared solution counted per completed challenge
3. The Thresholds are currently:
Platinum - 6
Gold - 4
Silver - 2
Bronze - 0

```sql
SELECT COUNT(SharedSolutionId) as TotalCompleted 
FROM (
SELECT DISTINCT SharedSolutionId 
FROM SharedSolution JOIN Challenge ON SharedSolution.challengeId = Challenge.Id 
JOIN Vote ON SharedSolution.Id = Vote.SharedSolutionId 
WHERE SharedSolution.userId = ${userId} AND rootType = 'SHAREDSOLUTION') unique_query
```

```ts

export const SharedBadgesFn = async (badges: AllBadges[], query: SharedTotals[]) => {
  const thresholds: { slug: SubmittedSolutions; threshold: number }[] = [
    { slug: 'platinum', threshold: 6 },
    { slug: 'gold', threshold: 4 },
    { slug: 'silver', threshold: 2 },
    { slug: 'bronze', threshold: 0 },
  ];
  const total = query?.[0]?.TotalCompleted || 0;
  const [highestBadge] = thresholds.filter((x) => total >= x.threshold);
  if (highestBadge) {
    badges.push({
      slug: highestBadge.slug,
      name: `Completed ${highestBadge.slug} Unique Solutions Badge`,
      shortName: `Shared`,
    });
  }
};

```

### Difficulty Completion Badge

Completed all the problems for a given level.

```sql
SELECT Difficulty, COUNT(Id) as TotalCompleted 
FROM (
SELECT DISTINCT Difficulty, Challenge.Id 
FROM Submission JOIN Challenge ON Submission.challengeId = Challenge.Id 
WHERE Submission.userId = ${userId} AND IsSuccessful = 1) unique_query 
GROUP BY Difficulty
```

logic:
```ts

export const DifficultyBadgesFn = async (badges: AllBadges[], query: Difficulty[]) => {
  const thresholds: { difficulty: DifficultyCompletion; threshold: number }[] = [
    { difficulty: 'BEGINNER', threshold: 1 },
    { difficulty: 'EASY', threshold: 13 },
    { difficulty: 'MEDIUM', threshold: 97 },
    { difficulty: 'HARD', threshold: 54 },
    { difficulty: 'EXTREME', threshold: 17 },
  ];
  query.forEach((currQuery) => {
    const levelThreshold = thresholds.find((x) => x.difficulty.toUpperCase() === currQuery.Difficulty);
    const completedAllChallenges = levelThreshold?.threshold === Number(currQuery.TotalCompleted);
    if (completedAllChallenges) {
      const pascalCase = `${currQuery.Difficulty[0]}${currQuery.Difficulty.substring(
        1,
      ).toLowerCase()}`;
      badges.push({
        slug: currQuery.Difficulty,
        name: `Completed ${pascalCase} Difficulty Badge`,
        shortName: currQuery.Difficulty?.toLowerCase(),
      });
    }
  });
};
```

### Similar Discussion/Current State
https://github.com/typehero/typehero/discussions/1304

**Comment by Hacksore** [comment]
the hub badges for ref 
https://github.com/Schweinepriester/github-profile-achievements

**Reply by CodeItQuick** [reply]
If you follow the link chain, I found this interface pretty helpful. Specifically changed "Details" to "History" as really to the user the Details are actually badge history.

![image](https://github.com/user-attachments/assets/bda78fcb-e0b7-444a-b3ac-5a366480fa4b)

**Reply by CodeItQuick** [reply]
Some pluralsight references as well:
### Streaks
![image](https://github.com/user-attachments/assets/7613a6c4-2ae4-4ac9-a546-64ce4a3cd7dc)
### Weekly and Annual Learning Goals
![image](https://github.com/user-attachments/assets/9a1de1eb-40d7-4d92-bd2e-9969adf3bcbd)

**Reply by CodeItQuick** [reply]
### Leet Code
![image](https://github.com/user-attachments/assets/8bc61f7c-fddc-447a-b7f5-1c39efd4ab7a)

Here are some of the badges you can earn on LeetCode and the requirements for each:

1.    Question Asker: Ask a question that receives at least 10 upvotes.
2.    Question Solver: Solve a question that has received at least 10 upvotes.
3.    Forum Supporter: Receive at least 10 upvotes on a comment or answer.
4.    Hot Topic: Start a thread that receives at least 100 views.
5.    Pundit: Post at least 10 comments or answers that receive at least 10 upvotes each.
6.    Problem Solver: Solve at least 100 problems on LeetCode.
7.    Challenger: Participate in at least one LeetCode coding challenge.
8.    Translation Contributor: Contribute a translation of a LeetCode problem description into another language.

**Comment by dca123** [comment]
Some achievements that are more community focused

- submit a challenge
- get a pinned solution
- Get X upvotes

How would you plan on implementing this ? Event driven or maybe a daily badge awarding process ?

**Reply by CodeItQuick** [reply]
It will be event driven yes. The first provided table are the "submission events". Everytime someone submits a challenge it'll send an event. Probably its own domain (not sure your familiarity with domain driven design, but that's my general definition of "domain" here).

I'm going to flush out the user achievements a bit now as well, since I think it will influence the design. I've thrown these 3 points into "user achievements" but I recognize they may actually fall into a different bucket.

**Comment by bautistaaa** [comment]
Where my head is currently at

For challenge based badges

- Badge for completing ALL easy / medium / hard / expert
- Badge for completing ALL 
- Sharing 10 different solutions (various thresholds)
- Commenting x number of times

We want to incentivize sharing and participating in the community.

Looking at the datamodel i am confused why we need a history table if `UserBadges` can be used. I'm assuming as badge is inserted into this table when they complete a badge. This table has a createdAt date by default and we just use that?

**Comment by CodeItQuick** [comment]
Discussion with boek:

![image](https://github.com/user-attachments/assets/a7ff30a8-46ca-4a82-b3eb-a05b3aeb07bc)

**Reply by bautistaaa** [reply]
thanks @boek !

**Reply by CodeItQuick** [reply]
Seconding the thanks boek. I'm going to work this diagram back into the design when I get time.

**Comment by CodeItQuick** [comment]
This is the original opening post, before I updated to incorporate beok's comment:


### Community Achievements
Prioritize community achievements and use thresholds for community stuff


### Similar Discussion/Current State
https://github.com/typehero/typehero/discussions/1304

### Open Questions
1. Do we send two events for a single submission, or just one event with the two pieces of data. Eg: Two events, one with Activity: Difficulty, ActivityValue: Easy, and another Activity: Track, ActivityValue: AOT 2023. Or do we just do the current proposal, One Event, with Difficulty and Track attached. The advantage of the first is a single interface for the microservice/domain/whatever this is to receive events. The advantage of the second is its more concise. Leaning towards a single interface currently.
2. How to ensure no duplicated events? Is that responsibility on the sender (not the receiver). The receiver has to recognize two SubmissionID Guids are the same, but that's it?
3. Is ActivityValue "correct". Name + Data both seem wrong.


### Type Hero Website User Achievements
1. Submitting Challenges
5. Creating a Track

### TrashDev Discord User Achievements
1. Moderator
3. Joining the Discord
6. Discord VC Regular
7. Gained Lore in Discord

### Type Hero Github Open Source
1. Contributing to Type Hero Open Source
2. Create Discussion in Type Hero Open Source
4. Maintainer of Type Hero Open Source

### Challenge Submission Achievements
1. Finishing a track
9. Starting a track
10. Finishing x number of beginner/intermediate/hard
11. Finishing x number of challenges (total)
12. Successful Challenge Completed
13. Unsuccessful Challenges Attempted
14. Number of challenges completed in a single month?
15. Streaks - completed 1 challenge every month this year.
16. Streaks - completed 1 challenge every day of a week
17. Top 5%/10%/20% on site (according to completed challenges)
18. Hit Weekly Goal x1, x5, x10, x20, x26 (half year), x52 (for a year) (maybe "Goal" is complete 3 challenges per period)
19. Hit Monthly Goal x1, x3, x6, x12 (maybe "Goal" is complete 5 challenges per period)

### Examples:  
0->4 Easy Challenges (Bronze)  
4->10 Medium Challenges (Silver)  
0->20 Challenges in January (Bronze)  

### User Achievement Events
| SubmissionID | Source | Activity | ActivityCategory  |  ActivityValue  | SubmissionDate | UserId   |   
|--------------|-------------|-------------|----------|-------------|----------------|----------------|
| 1235-abcd    |  Website |    Community Contribution  | Submitted Challenge    |  23  | 01-01-2024     | 10291643 |   
| 1235-abce    | Discord | Role Change          | Moderator         |  dota2attitude |  01-01-2024     | 10291643 |   
| 1235-abcf    |  Github |  Pull Request          | Merged                  | CodeItQuick |  01-02-2024     | 10291643 |   
| 1235-abcg    | Github | Discussion          | Opened      |  CodeItQuick   |  01-02-2024     | 10291643 |   

### Option 1: Submission Achievement Events
| SubmissionID | Source | Activity | ActivityCategory    |  ActivityValue  |  SubmissionDate | UserId   |   
|--------------|-------------|-------------|----------|-------------|----------------|----------------|
| 1234-abcd    |  Website  |   Completed Difficulty      | Easy  |  13  |  01-01-2024     | 10291643 |   
| 1234-abce    |  Website  |   Completed Track    | AOT 2023  |  13    | 01-01-2024     | 10291643 |   
| 1234-abcf    |  Website  |   Completed Track     | AOT 2024  |  15 |   01-01-2024     | 10291643 |   
| 1234-abcg    |  Website  |   Started Difficulty  | Medium  | 15  |  01-01-2024     | 10291643 |   

### Option 2: Submission Achievement Events
| SubmissionID | ChallengeId | Difficulty | Track    | IsCompleted | SubmissionDate | UserId   |   
|--------------|-------------|------------|----------|-------------|----------------|----------
| 1234-abcd    | 13          | Easy       | AOT 2023 | True        | 01-01-2024     | 10291643 |   
| 1234-abce    | 12          | Easy       | AOT 2023 | True        | 01-01-2024     | 10291643 |   
| 1234-abcf    | 14          | Easy       | AOT 2023 | True        | 01-02-2024     | 10291643 |   
| 1234-abcg    | 15          | Medium     | AOT 2024 | True        | 01-02-2024     | 10291643 |   


### Badges
| Id  | BadgeCategory      | CriteriaColumn        | CriteriaValue | Level    |  Threshold
|-----|--------------------|-----------------------|---------------|----------|--------|
| 13  |  AOT 2023           | Track                 | AOT 2023      | Bronze   |   10 |
| 14  |  AOT 2023           | Track                 | AOT 2023      | Silver   | 20 |
| 15  |  AOT 2023           | Track                 | AOT 2023      | Gold     | 30 |
| 16  |  AOT 2023           | Track                 | AOT 2023      | Platinum |  40 |
| 17  |  AOT 2024           | Track                 | AOT 2024      | Gold     |  30 |
| 18  | Easy Difficulty Badge     | Difficulty            | Easy          | Bronze   |  10 |
| 19  | Challenge Submittor | Submitted Challenge  | 10      | Silver  | 20 |
| 20  | Discord Moderator | SubmissionDate  | 7 Days In Row      | Diamond  |  60 |
| 21  | Discord Lore Contributor | GainedLore  | 5      | Bronze  |  10 |
| 22  | Type Hero Open Source Contributor | ClosedPR  | 5      | ContributionStarter  |  10 |
* 13->19 are SubmissionAchievements
* 20->22 are UserAchievements, and I may need another column to grab the "source" of the UserAchievement (eg: Discord, Github, Act of Trash, etc.)

### UserBadges
| Id | UserId         | BadgeId | AchievementDate  | 
|----|----------------|---------|------------------|
| 7  | 10291643       | 13      | January 10, 2024 |
| 8  | 10291643       | 17      | January 10, 2024 | 
| 9  | 10291643       | 18      | January 12, 2024 |
| 10 | 10291643       | 19      | February 7, 2024 |

### BadgeHistory
| EventId | ChallengeId | UserBadgeId | Criteria | CriteriaValue    | AchievementDate  | 
|---------|-------------|---------------------|------------|----------|------------------|
| 1       | 13          | 7                   | Track       | AOT 2023 | January 10, 2024 |
| 2       | 12          | 7                   | Track      | AOT 2023 | January 10, 2024 | 
| 3       | 15          | 7                   | Track     | AOT 2023 | January 12, 2024 |
| 4       | 13          | 10                   | Difficulty     | Easy | January 10, 2024 |
| 5       | 12          | 10                   | Difficulty     | Easy | January 10, 2024 |
| 6       | 18          | 8                   | Track       | AOT 2024 | February 7, 2024 |
